.github/workflows/build.yml

```diff
@@ -29,24 +29,38 @@ jobs:
         id: checkout
         uses: actions/checkout@v3
 
-      - name: Prepare
+      - name: Update system packages
         id: prepare
         run: |
-          # Install test tools.
-          if apt-get update && apt-get install -y git python3-venv; then
-             # Debian-specific temporary code:
-             # Installation of python3-venv can be removed as soon as a
-             # base image with a release including #33822 is available
-             :
-          else
-             export PATH="build/bin:$PATH"
-             eval $(sage-print-system-package-command auto update)
-             eval $(sage-print-system-package-command auto --spkg --yes --no-install-recommends install git)
-          fi
+          export PATH="build/bin:$PATH"
+          eval $(sage-print-system-package-command auto update)
+          eval $(sage-print-system-package-command auto --spkg --yes --no-install-recommends install git)
+
+      - name: Incremental build and test
```

Comment:
Why is this coming before the actual build/is this not building things twice now?

Reply:
We could, in fact, also run the main build on the incrementally updated source tree.

Reply:
In this "incremental" build, what is build if one changes a python file, a cython file and say something in the build infrastructure (say a package update and a change in setup.py)?

Reply:
The git acrobatics has exactly the same effect as doing a "git checkout" to the new ref in a full build of the Sage distribution on a developer's computer: The timestamps (mtime) of unmodified files are kept the same, and the timestamps of modified files are newer than any file that was there before.
It is looking so complicated because the source tree in our Docker images is actually not a git working copy. This is why I have to turn it surgically into a working copy (without messing up the timestamps).

Running `make build` results in an incremental build: For spkgs, if the declared package version is changed, then (because they no longer match the installation records in `{local,venv}/var/libs/sage/installed` these packages are reinstalled.
For the Sage library, the normal in-tree editable build is run, which is incremental (only the modified Cython files are cythonized and the result compiled). Because we are using editable mode, nothing interesting is happening with the modified Python files.

Reply:
Thanks for the details!

But doesn't this mean that sometimes the incremental build (+ test) fails? At least locally, I sometimes have to hard-reset everything and run the whole bootstrap + configure + make cycle. In this cases, the workflow would fail, right? I'm not sure if this then more confusing than helpful.

Reply:
By "hard reset", do you mean that you do "make distclean" (= remove local and venv)?

Reply:
In my experience, incremental builds are very robust, with the following exceptions:
- the docbuild process
- after indiscriminate updates of system packages (we don't do that here on this workflow)

Reply:
> By "hard reset", do you mean that you do "make distclean" (= remove local and venv)?

Sometimes, but most often I only have to run bootstrap & configure again.

Reply:
> In my experience, incremental builds are very robust, with the following exceptions:

Okay, then I let someone else with more experience on the build process have a look at this PR.

If the incremental build is not (yet) reliable enough, I would favor a normal build + test on changed files + long test; and/or extracting the incremental build to a new workflow.

Reply:
> > By "hard reset", do you mean that you do "make distclean" (= remove local and venv)?
> 
> Sometimes, but most often I only have to run bootstrap & configure again.

Ah OK. We can certainly run `bootstrap` manually before doing the incremental build. I'll make this change.
(configure definitely does not have to be run manually.)

Reply:
> If the incremental build is not (yet) reliable enough, I would favor a normal build + test on changed files + long test; and/or extracting the incremental build to a new workflow.

@tobiasdiez @kwankyu I have now implemented a fallback to non-incremental when the incremental build fails – both for build and for docbuild.